File: server/src/main/java/com/example/server/service/NotificationService.java
package com.example.server.service;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.server.entity.Notification;
import com.example.server.entity.User;
import com.example.server.repository.NotificationRepository;

@Service
public class NotificationService {

    private static final Logger log = LoggerFactory.getLogger(NotificationService.class);

    private final NotificationRepository notificationRepository;

    // Constructor
    public NotificationService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    public List<Notification> getUserNotifications(String userId) {
        log.info("Fetching notifications for user: {}", userId);
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }

    public List<Notification> getUnreadNotifications(String userId) {
        log.info("Fetching unread notifications for user: {}", userId);
        return notificationRepository.findByUserIdAndIsReadFalseOrderByCreatedAtDesc(userId, false);
    }

    public long getUnreadCount(String userId) {
        return notificationRepository.countByUserIdAndIsReadFalse(userId, false);
    }

    @Transactional
    public void markAsRead(String notificationId) {
        log.info("Marking notification as read: {}", notificationId);
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Notification not found"));
        notification.setIsRead(true);
        notificationRepository.save(notification);
    }

    @Transactional
    public void markAllAsRead(String userId) {
        log.info("Marking all notifications as read for user: {}", userId);
        List<Notification> notifications = notificationRepository.findByUserIdAndIsReadFalseOrderByCreatedAtDesc(userId, false);
        notifications.forEach(n -> n.setIsRead(true));
        notificationRepository.saveAll(notifications);
    }

    @Transactional
    public Notification createNotification(String userId, User user, String type, String title, String message, String relatedEntityId) {
        Notification notification = new Notification();
        notification.setUserId(userId);
        notification.setUser(user);
        notification.setType(type);
        notification.setTitle(title);
        notification.setMessage(message);
        notification.setRelatedEntityId(relatedEntityId);
        return notificationRepository.save(notification);
    }
}

File: client/src/services/notificationService.js
import apiClient from './apiClient';

function unwrapList(response) {
  if (Array.isArray(response)) return response;
  if (Array.isArray(response?.data)) return response.data;
  return [];
}

// Notifications Service
export const notificationService = {
  async getUserNotifications(userId) {
    try {
      const response = await apiClient.get(`/notifications?userId=${userId}`);
      return {
        ok: true,
        data: unwrapList(response),
        unreadCount: response.unreadCount || 0,
      };
    } catch (error) {
      return {
        ok: false,
        message: error.message,
        data: [],
        unreadCount: 0,
      };
    }
  },

  async getUnreadNotifications(userId) {
    try {
      const response = await apiClient.get(`/notifications/unread?userId=${userId}`);
      return {
        ok: true,
        data: unwrapList(response),
      };
    } catch (error) {
      return {
        ok: false,
        message: error.message,
        data: [],
      };
    }
  },

  async markAsRead(notificationId) {
    try {
      const response = await apiClient.patch(
        `/notifications/${notificationId}/read`,
        {}
      );
      return {
        ok: true,
        data: response.data,
      };
    } catch (error) {
      return {
        ok: false,
        message: error.message,
      };
    }
  },

  async markAllAsRead(userId) {
    try {
      const response = await apiClient.patch(
        `/notifications/read-all?userId=${userId}`,
        {}
      );
      return {
        ok: true,
        data: response.data,
      };
    } catch (error) {
      return {
        ok: false,
        message: error.message,
      };
    }
  },
};

export default notificationService;
